src/desktop/sse/SecretStorage.ts

```diff
@@ -33,8 +94,122 @@ export class KeytarSecretStorage implements SecretStorage {
 	}
 
 	setPassword(service: string, account: string, password: string): Promise<void> {
-		const newOp = this.lastOp.catch(noOp).then(() => setPassword(service, account, password))
+		const newOp = this.lastOp.catch(noOp).then(() => this._setPassword(service, account, password))
 		this.lastOp = newOp
 		return newOp
 	}
 }
+
+/**
+ * Secret Storage impl using the electron 15+ SafeStorage API
+ *
+ * Note: the main thread will be blocked while the keychain is being unlocked,
+ * potentially for as long as the user takes to enter a password.
+ * We're asking for access before any windows are created, which should prevent
+ * any weirdness arising from that.
+ */
+export class SafeStorageSecretStorage implements SecretStorage {
+	private initialized = false
+
+	constructor(
+		private readonly electron: typeof Electron.CrossProcessExports,
+		private readonly fs: typeof FsModule,
+		private readonly path: typeof PathModule,
+		private readonly keytarSecretStorage: SecretStorage,
+	) {}
+
+	async getPassword(service: string, account: string): Promise<string | null> {
+		await this.assertAvailable()
+		const keyPath = this.getKeyPath(service, account)
+		try {
+			const encPwBuffer = await this.fs.promises.readFile(keyPath)
+			return this.electron.safeStorage.decryptString(encPwBuffer)
+		} catch (e) {
+			if (e.code === "ENOENT") {
+				// we might not have the key in safeStorage yet, but we might have it in the keytar storage.
+				return await this.migrateKeytarPassword(service, account)
+			}
+			throw e
+		}
+	}
+
+	async setPassword(service: string, account: string, password: string): Promise<void> {
+		await this.assertAvailable()
+		const keyPath = this.getKeyPath(service, account)
+		const cypherBuffer = this.electron.safeStorage.encryptString(password)
+		return this.fs.promises.writeFile(keyPath, cypherBuffer)
+	}
+
+	private getKeyPath(service: string, account: string): string {
+		const fname = service.concat("-", account)
+		const safeStoragePath = this.getSafeStoragePath()
+		return this.path.join(safeStoragePath, fname)
+	}
+
+	/**
+	 * this should always be a path inside the user's home directory (or equivalent)
+	 * @private
+	 */
+	private getSafeStoragePath(): string {
+		return this.path.join(this.electron.app.getPath("userData"), "safe_storage")
+	}
+
+	/**
+	 * ensures that the safe_storage directory exists and that we can use the
+	 * safeStorage API
+	 * @private
+	 */
+	private async assertAvailable(): Promise<void> {
+		await this.electron.app.whenReady()
+		await this.fs.promises.mkdir(this.getSafeStoragePath(), { recursive: true })
+		if (this.electron.safeStorage.isEncryptionAvailable()) {
+			if (!this.initialized && process.platform === "linux") {
```

Comment:
you could use some kind of "once" thing like `lazyMemoized` to not print every time

Reply:
isn't that functionally the same thing

Reply:
yup, just saying